app/uk/gov/hmrc/gform/formtemplate/JsonSchemeValidator.scala

```diff
@@ -67,59 +43,193 @@ object JsonSchemeValidator {
+    error.keyword match {
+      case "pattern" =>
+        Some(
+          (
+            error.location.substring(0, error.location.lastIndexOf("/")),
+            splitErrorLocation(splitErrorLocation.length - 3)
+          )
+        )
+      case "required" =>
+        Some((error.location, splitErrorLocation.last))
+      case _ => None
+    }
+
+  private def tryConvertErrorLocationToId(json: Json, location: String): String = {
+    val errorLocationSections = location.split("/").toList.tail
+
+    getErrorLocationId(json, errorLocationSections) match {
+      // If cannot get ID of location, use original location message instead
+      case None => location
+
+      case Some(errorLocationId: Json) =>
+        errorLocationId.asString match {
+          case None                                => location
+          case Some(errorLocationIdString: String) => errorLocationIdString
+        }
+    }
+  }
+
+  @tailrec
+  private def getErrorLocationId(json: Json, remainingSections: List[String]): Option[Json] =
+    remainingSections.headOption match {
+      // No more sections to traverse, so get ID of current section to return
+      case None =>
+        json.hcursor.downField("id").as[Json] match {
+          case Left(_)                      => None
+          case Right(errorLocationId: Json) => Some(errorLocationId)
+        }
+
+      // More sections to traverse, so check if current json is a List of Json or a Json
+      case Some(section: String) =>
+        json.as[List[Json]] match {
+
+          // If Json, go to next section
+          case Left(_) =>
+            json.hcursor.downField(section).as[Json] match {
+              case Left(_)                      => None
+              case Right(nextJsonSection: Json) => getErrorLocationId(nextJsonSection, remainingSections.tail)
+            }
+
+          // If List of Json, get Int of next section and index the List to go to next section
+          case Right(jsonList: List[Json]) =>
+            section.toIntOption match {
+              case Some(sectionInt: Int) => getErrorLocationId(jsonList(sectionInt), remainingSections.tail)
+              case None                  => None
+            }
+        }
+    }
+
+  private def propertyRequirementsFromSchema(schema: Json, property: String): Result[Map[String, Map[String, String]]] =
+    schema.hcursor
+      .downField("$defs")
+      .downField("fields")
+      .downField("dependencies")
+      .downField(property)
+      .downField("properties")
+      .as[Map[String, Map[String, String]]]
+
+  private def requiredPropertyValuesFromPattern(requiredPropertyAndPattern: Map[String, Map[String, String]]): String =
+    requiredPropertyAndPattern
+      .map { case (requiredProperty, pattern) =>
+        requiredProperty -> pattern
+          .map { case (_, requiredValues) =>
+            requiredValues.substring(1, requiredValues.length - 1).replace("|", ", ")
+          }
+          .mkString(", ")
+      }
```

Comment:
map over/use only values from the HashMap if possible: 
```suggestion
    requiredPropertyAndPattern.view
      .mapValues { pattern =>
        pattern.values
          .map { requiredValues =>
            requiredValues.substring(1, requiredValues.length - 1).replace("|", ", ")
          }
          .mkString(", ")
      }
```